docs/blog/posts/sled-integration.md

```diff
@@ -0,0 +1,248 @@
+    // populated after load, randomly generated prior to load.
+    id: u32,
+
+    // program_bytes is used to temporarily cache the raw program data during
+    // the loading process.  It MUST be cleared following a load so that there
+    // is not a long lived copy of the program data living on the heap.
+    program_bytes: Vec<u8>,
+}
+```
+
+All of the fields are now removed in favor of a private reference to the unique
+[`sled::Tree`](https://docs.rs/sled/latest/sled/struct.Tree.html) instance for
+this `ProgramData` which is named using the unique kernel id for the program.
+Each `sled::Tree` represents a single logical keyspace / namespace / bucket which
+allows key generation to be kept simple, i.e `db.get("program_<ID>_name")`
+now can be `db_tree_prog_0000.get("program_name)`. Additionally getters and setters are now
+built for each existing field so that access to the db can be controlled
+and the existing degree of type safety can be maintained:
+
+```rust
+...
+pub(crate) fn set_name(&mut self, name: &str) -> Result<(), BpfmanError> {
+    self.insert("name", name.as_bytes())
+}
+
+pub(crate) fn get_name(&self) -> Result<String, BpfmanError> {
+    self.get("name").map(|v| bytes_to_string(&v))
+}
+...
+```
+
+Therefore, `ProgramData` is now less of a container for program data and more of a
+wrapper for accessing program data.  The getters/setters act as a bridge
+between standard Rust types and the raw bytes stored in the database,
+i.e the [`sled::IVec` type](https://docs.rs/sled/latest/sled/struct.IVec.html).
+
+Once this was completed for all the relevant fields on all the relevant types,
+[see pull request #874](https://github.com/bpfman/bpfman/pull/874),the
+data bpfman needed for it's managed eBPF programs was now automatically synced
+to disk :partying_face:
+
+## Tradeoffs
+
+All design changes come with some tradeoffs...for bpfman's conversion to using
+sled the main negative ended up being with the complexity introduced with the [`sled::IVec` type](https://docs.rs/sled/latest/sled/struct.IVec.html).
+It is basically just a thread-safe reference-counting pointer to a raw byte slice,
+and the only type raw database operations can be performed with.  Previously when
+using `serde_json` all serialization/deserialization was automatically handled,
+however with sled the conversion is manual handled internally.  Therefore, instead
+of a library handling the conversion of a rust string (`std::string::String`) to
+raw bytes `&[u8]` bpfman has to handle it internally, using [`std::string::String::as_bytes`](https://doc.rust-lang.org/std/string/struct.String.html#method.as_bytes) and `bpfman::utils::bytes_to_string`:
+
+```rust
+pub(crate) fn bytes_to_string(bytes: &[u8]) -> String {
+    String::from_utf8(bytes.to_vec()).expect("failed to convert &[u8] to string")
+}
+```
+
+For strings conversion is simple enough, but when working with more complex rust
+types like `Hashmaps` and `Vectors` this became a bit more of a headache.
```

Comment:
Did you want to talk about flattening these types here - i.e how hashmaps could be db trees and how you might encode arrays?

Reply:
Took a stab here, mostly just explained how we did it